### ISSUE TITLE
Incompatibility with NDS.Live 2024.03

### ISSUE THREAD

**Issue by jo-me** [woven-planet/go-zserio#131]
Hi, 
the 2024.03 schema version seems to be incompatible with the current go-zserio implementation.

The NDS.Live package builder forces a zserio 2.13 version with this new schema so I don't know which change actually causes the generator issues.

So here is the log when running against the display layer.
When using the previous versions (2023.12 + zserio 2.12) none of the issues below are reported.
The fatal error regarding the layer.zs also appears when generating other layer.zs.

```

2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/all.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/_module.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/attributemap.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/color.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/conditions.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/geometry.zs...
line 306:2 reportAttemptingFullContext d=60 (expression), input='>longitude;/** Latitude. */int<(31-shift)+1>latitude;/** Shift longitude value back to the left. */function'
line 300:31 reportContextSensitivity d=60 (expression), input='>longitude;'
line 306:2 reportAttemptingFullContext d=60 (expression), input='>latitude;/** Shift longitude value back to the left. */function'
line 303:30 reportContextSensitivity d=60 (expression), input='>latitude;'
line 331:2 reportAttemptingFullContext d=60 (expression), input='>deltaLongitude;/** Latitudinal delta to base geographic position. */int<numBits+1>deltaLatitude;/** Shift longitude value back to the left. */function'
line 325:33 reportContextSensitivity d=60 (expression), input='>deltaLongitude;'
line 331:2 reportAttemptingFullContext d=60 (expression), input='>deltaLatitude;/** Shift longitude value back to the left. */function'
line 328:32 reportContextSensitivity d=60 (expression), input='>deltaLatitude;'
line 359:2 reportAttemptingFullContext d=60 (expression), input='>longitude;/** Latitude. */int<(31-shiftXY)+1>latitude;/** Elevation in cm above/below WGS 84 reference ellipsoid. */int<31-shiftZ+1>elevation;/** Shift longitude value back to the left. */function'
line 350:33 reportContextSensitivity d=60 (expression), input='>longitude;'
line 359:2 reportAttemptingFullContext d=60 (expression), input='>latitude;/** Elevation in cm above/below WGS 84 reference ellipsoid. */int<31-shiftZ+1>elevation;/** Shift longitude value back to the left. */function'
line 353:32 reportContextSensitivity d=60 (expression), input='>latitude;'
line 359:2 reportAttemptingFullContext d=60 (expression), input='>elevation;/** Shift longitude value back to the left. */function'
line 356:30 reportContextSensitivity d=60 (expression), input='>elevation;'
line 393:2 reportAttemptingFullContext d=60 (expression), input='>deltaLongitude;/** Latitudinal delta to base geographic position. */int<numBitsXY+1>deltaLatitude;/** Elevation delta to base geographic position. */int<numBitsZ+1>deltaElevation;/** Shift longitude value back to the left. */function'
line 384:35 reportContextSensitivity d=60 (expression), input='>deltaLongitude;'
line 393:2 reportAttemptingFullContext d=60 (expression), input='>deltaLatitude;/** Elevation delta to base geographic position. */int<numBitsZ+1>deltaElevation;/** Shift longitude value back to the left. */function'
line 387:34 reportContextSensitivity d=60 (expression), input='>deltaLatitude;'
line 393:2 reportAttemptingFullContext d=60 (expression), input='>deltaElevation;/** Shift longitude value back to the left. */function'
line 390:34 reportContextSensitivity d=60 (expression), input='>deltaElevation;'
line 550:1 reportAttemptingFullContext d=60 (expression), input='>start;};'
line 549:25 reportContextSensitivity d=60 (expression), input='>'
line 1133:1 reportAttemptingFullContext d=60 (expression), input='>index:index<positionBufferSize;};'
line 1132:41 reportContextSensitivity d=60 (expression), input='>index:'
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/grid.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/icons.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/language.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/location.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/packaging.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/properties.zs...
line 63:2 token recognition error at: '@r'
line 63:11 extraneous input 'METRIC_UNIT' expecting {',', '}'}
line 68:2 token recognition error at: '@r'
line 68:11 extraneous input 'IMPERIAL_UNIT' expecting {',', '}'}
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/speech.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/types.zs...
line 282:1 reportAttemptingFullContext d=60 (expression), input='>minScaleDenominator;};'
line 281:61 reportContextSensitivity d=60 (expression), input='>'
line 422:27 reportAttemptingFullContext d=60 (expression), input='>startPosition;/** End percentage position on the feature.*/bit<numBits>endPosition:'
line 419:28 reportContextSensitivity d=60 (expression), input='>startPosition;'
line 424:2 reportAttemptingFullContext d=60 (expression), input='>endPosition:endPosition>startPosition;function'
line 422:27 reportContextSensitivity d=60 (expression), input='>endPosition:'
line 424:2 reportAttemptingFullContext d=60 (expression), input='>startPosition;function'
line 422:41 reportContextSensitivity d=60 (expression), input='>'
line 444:2 reportAttemptingFullContext d=60 (expression), input='>position;function'
line 442:23 reportContextSensitivity d=60 (expression), input='>position;'
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/vehicle.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/display/_module.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/display/instantiations.zs...
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/display/layer.zs...
panic: interface conversion: interface {} is nil, not *ast.Expression

goroutine 1 [running]:
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitFieldOptionalClause(0xc0006c7128?, 0x3e?)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:233 +0x69
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitStructureFieldDefinition(0xc0003d8c30?, 0xc0006c7128)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:201 +0x1d3
github.com/woven-planet/go-zserio/internal/parser.(*StructureFieldDefinitionContext).Accept(0x8a2800?, {0xa29158?, 0xc00049a768})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/parser/zserio_parser.go:3015 +0x99
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).Visit(0xc0006c6ff0?, {0x7f043c2c2a38?, 0xc0006c7128?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:28 +0x114
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitStructureDeclaration(0x4?, 0xc0006c6ff0)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:174 +0x29a
github.com/woven-planet/go-zserio/internal/parser.(*StructureDeclarationContext).Accept(0x10?, {0xa29158?, 0xc00049a768})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/parser/zserio_parser.go:2769 +0x99
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).Visit(0xc0003072d0?, {0x7f043c2c2968?, 0xc0006c6ff0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:28 +0x114
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitChildren(0x9501e0?, {0xa2fef0?, 0xc0006c6fd8?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:38 +0xd2
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitTypeDeclaration(0xc000643818?, 0xa29158?)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:134 +0x1d
github.com/woven-planet/go-zserio/internal/parser.(*TypeDeclarationContext).Accept(0x10?, {0xa29158?, 0xc00049a768})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/parser/zserio_parser.go:1714 +0x99
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).Visit(0xc0003072e0?, {0x7f043c2c2900?, 0xc0006c6fd8?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:28 +0x114
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitChildren(0x9501e0?, {0xa2ff68?, 0xc0006c6fc0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:38 +0xd2
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitLanguageDirective(0xc0001ba1e0?, 0xa29158?)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:129 +0x1d
github.com/woven-planet/go-zserio/internal/parser.(*LanguageDirectiveContext).Accept(0xd0?, {0xa29158?, 0xc00049a768})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/parser/zserio_parser.go:1482 +0x99
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).Visit(0x8cdb60?, {0x7f043c3289d8?, 0xc0006c6fc0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:28 +0x114
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitPackageDeclaration(0xc00014fd40?, 0xc00070b0b0)
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:59 +0x44a
github.com/woven-planet/go-zserio/internal/parser.(*PackageDeclarationContext).Accept(0xc00070b0b0?, {0xa29158?, 0xc00049a768})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/parser/zserio_parser.go:783 +0x99
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).Visit(0xc00078c7c0?, {0x7f043dc1f158?, 0xc00070b0b0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/visitor/visitor.go:28 +0x114
github.com/woven-planet/go-zserio/internal/model.PackageFromFile({0xc00003fb80?, 0x0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/model/package.go:23 +0x209
github.com/woven-planet/go-zserio/internal/model.(*Model).parseFile(0xc00037a2c0, {0xc00003fb80, 0x34})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/model/model.go:89 +0xc5
github.com/woven-planet/go-zserio/internal/model.FromFiles({0xc000378800, 0x2e, 0x8cd740?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/model/model.go:28 +0x9e
github.com/woven-planet/go-zserio/internal/model.FromFilesystem({0xc000256700?, 0xc00010bee0?, 0x0?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/model/model.go:54 +0x4d
github.com/woven-planet/go-zserio/internal/cli.glob..func1(0xc00035aa00?, {0xc000256700?, 0x4?, 0x95b7cc?})
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/cli/generate.go:28 +0x30
github.com/spf13/cobra.(*Command).execute(0xe18ae0, {0xc000256690, 0x7, 0x7})
        /home/jochen/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xe18d60)
        /home/jochen/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x38d
github.com/spf13/cobra.(*Command).Execute(...)
        /home/jochen/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
github.com/woven-planet/go-zserio/internal/cli.Execute()
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/internal/cli/root.go:15 +0x1a
main.main()
        /home/jochen/go/pkg/mod/github.com/woven-planet/go-zserio@v0.7.0/cmd/go-zserio/main.go:8 +0xf
```

Jochen Mehlhorn <jochen.mehlhorn@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH

[Provider Information](https://github.com/Daimler/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by jo-me** [comment]
```
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/core/properties.zs...
line 63:2 token recognition error at: '@r'
line 63:11 extraneous input 'METRIC_UNIT' expecting {',', '}'}
line 68:2 token recognition error at: '@r'
line 68:11 extraneous input 'IMPERIAL_UNIT' expecting {',', '}'}
```

seems to be related to the introduction of the `@removed` token:
https://github.com/ndsev/zserio/issues/486

**Comment by jo-me** [comment]
```
2024/04/29 11:15:01 Parsing file ndslive-display-2024.03/zs/display/layer.zs...
panic: interface conversion: interface {} is nil, not *ast.Expression

goroutine 1 [running]:
github.com/woven-planet/go-zserio/internal/visitor.(*Visitor).VisitFieldOptionalClause(0xc0006c7128?, 0x3e?)
```

is apparently caused by the use of the isset() operator in place of plain logic throughout all layer.zs and metadata.zs files.

![image](https://github.com/woven-planet/go-zserio/assets/10351272/e9aa3b09-ed03-489c-bdc9-969815ed8c12)


Seems like it is not implemented in this lib and therefore not a recognized expression. I would have suspected it to be implemented here, right?  https://github.com/woven-planet/go-zserio/blob/main/internal/ast/expression.go

**Comment by wichert** [comment]
Hi Jochem,

this week is Golden Week in Japan, so it might take a little bit of time
before we can take a look at this. Your analysis looks spot-on.

**Comment by Danaozhong** [comment]
@jo-me Thank you so much again for your bug report! I can confirm that the issue you found is caused by two issues:
- `go-zserio` was missing support for the `isset()` operator, which checks if bits within a bitmask are set.
- The currently used antlr definitions on `main` don't support the `deprecated`, `deleted` [keywords](https://github.com/ndsev/zserio/issues/486).

I raised a small PR to resolve these issues. https://github.com/woven-planet/go-zserio/pull/132 With these changes, I was able to compile the `display` module successfully. Do you mind to give it a try?

Thank you again.

**Comment by jo-me** [comment]
Thanks @Danaozhong  for getting back to this so quickly!
I've tried it out and it completed the code generation without any crashes. 🎉

There was some log output in between where i don't know whether those are just informative or indicate some problem. 
Everything starting with "line ..."

What do you think? 

```
2024/05/07 18:15:58 Parsing file display-2024.03/zs/all.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/_module.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/attributes.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/instantiations.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/layer.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/metadata.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/properties.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/characteristics/types.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/_module.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/attributemap.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/color.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/conditions.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/geometry.zs...
line 306:2 reportAttemptingFullContext d=62 (expression), input='>longitude;/** Latitude. */int<(31-shift)+1>latitude;/** Shift longitude value back to the left. */function'
line 300:31 reportContextSensitivity d=62 (expression), input='>longitude;'
line 306:2 reportAttemptingFullContext d=62 (expression), input='>latitude;/** Shift longitude value back to the left. */function'
line 303:30 reportContextSensitivity d=62 (expression), input='>latitude;'
line 331:2 reportAttemptingFullContext d=62 (expression), input='>deltaLongitude;/** Latitudinal delta to base geographic position. */int<numBits+1>deltaLatitude;/** Shift longitude value back to the left. */function'
line 325:33 reportContextSensitivity d=62 (expression), input='>deltaLongitude;'
line 331:2 reportAttemptingFullContext d=62 (expression), input='>deltaLatitude;/** Shift longitude value back to the left. */function'
line 328:32 reportContextSensitivity d=62 (expression), input='>deltaLatitude;'
line 359:2 reportAttemptingFullContext d=62 (expression), input='>longitude;/** Latitude. */int<(31-shiftXY)+1>latitude;/** Elevation in cm above/below WGS 84 reference ellipsoid. */int<31-shiftZ+1>elevation;/** Shift longitude value back to the left. */function'
line 350:33 reportContextSensitivity d=62 (expression), input='>longitude;'
line 359:2 reportAttemptingFullContext d=62 (expression), input='>latitude;/** Elevation in cm above/below WGS 84 reference ellipsoid. */int<31-shiftZ+1>elevation;/** Shift longitude value back to the left. */function'
line 353:32 reportContextSensitivity d=62 (expression), input='>latitude;'
line 359:2 reportAttemptingFullContext d=62 (expression), input='>elevation;/** Shift longitude value back to the left. */function'
line 356:30 reportContextSensitivity d=62 (expression), input='>elevation;'
line 393:2 reportAttemptingFullContext d=62 (expression), input='>deltaLongitude;/** Latitudinal delta to base geographic position. */int<numBitsXY+1>deltaLatitude;/** Elevation delta to base geographic position. */int<numBitsZ+1>deltaElevation;/** Shift longitude value back to the left. */function'
line 384:35 reportContextSensitivity d=62 (expression), input='>deltaLongitude;'
line 393:2 reportAttemptingFullContext d=62 (expression), input='>deltaLatitude;/** Elevation delta to base geographic position. */int<numBitsZ+1>deltaElevation;/** Shift longitude value back to the left. */function'
line 387:34 reportContextSensitivity d=62 (expression), input='>deltaLatitude;'
line 393:2 reportAttemptingFullContext d=62 (expression), input='>deltaElevation;/** Shift longitude value back to the left. */function'
line 390:34 reportContextSensitivity d=62 (expression), input='>deltaElevation;'
line 550:1 reportAttemptingFullContext d=62 (expression), input='>start;};'
line 549:25 reportContextSensitivity d=62 (expression), input='>'
line 1133:1 reportAttemptingFullContext d=62 (expression), input='>index:index<positionBufferSize;};'
line 1132:41 reportContextSensitivity d=62 (expression), input='>index:'
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/grid.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/icons.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/language.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/location.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/packaging.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/properties.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/speech.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/types.zs...
line 282:1 reportAttemptingFullContext d=62 (expression), input='>minScaleDenominator;};'
line 281:61 reportContextSensitivity d=62 (expression), input='>'
line 422:27 reportAttemptingFullContext d=62 (expression), input='>startPosition;/** End percentage position on the feature.*/bit<numBits>endPosition:'
line 419:28 reportContextSensitivity d=62 (expression), input='>startPosition;'
line 424:2 reportAttemptingFullContext d=62 (expression), input='>endPosition:endPosition>startPosition;function'
line 422:27 reportContextSensitivity d=62 (expression), input='>endPosition:'
line 424:2 reportAttemptingFullContext d=62 (expression), input='>startPosition;function'
line 422:41 reportContextSensitivity d=62 (expression), input='>'
line 444:2 reportAttemptingFullContext d=62 (expression), input='>position;function'
line 442:23 reportContextSensitivity d=62 (expression), input='>position;'
2024/05/07 18:15:58 Parsing file display-2024.03/zs/core/vehicle.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/_module.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/instantiations.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/layer.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/metadata.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/reference/_module.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/reference/types.zs...
2024/05/07 18:15:58 Parsing file display-2024.03/zs/display/types.zs...
line 3405:2 reportAttemptingFullContext d=62 (expression), input='>elevation;/** Normal vector. */NormalSphere'
line 3402:30 reportContextSensitivity d=62 (expression), input='>elevation;'
```

**Comment by Danaozhong** [comment]
`reportAttemptingFullContext` and `reportContextSensitivity` are callbacks from the antlr parser reporting an ambiguity in the zserio code parsing. These are not syntax errors, and as such the parse tree will be generated correctly.
Since these antlr parser files are the same as the reference `zserio` implementation, I think we can safely ignore them. I will raise a separate issue for this. Perhaps we can suppress the antlr parser ambiguity resolution.

**Comment by jo-me** [comment]
Excellent, thanks for the explanation. 
And thanks again for the effort to keep this lib up to date 👍

**Comment by Danaozhong** [comment]
@jo-me I merged the PR, and created a new release of `go-zserio` (v0.8.0). If your problem is solved, could you please close this issue? Thanks!
